### PR TITLE
GH-4891 LMDB store: Fix computation of next ID and GC logic

### PR DESCRIPTION
**Pull request by kenwenzel**
GitHub issue resolved: #4891 
Briefly describe the changes proposed in this PR:

Correctly load the next ID to avoid collisions with already assigned IDs.
Use reference counting for data types of literals and namespaces of URIs to
ensure referenced values are not GCed and are fully deleted if not referenced anymore.

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [x] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [x] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

### CODE REVIEW

**Comment by kenwenzel** [comment]
@hmottestad This is ready for review. Maybe I could squash the commits a bit...

File: core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/LmdbContextIterator.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+			throw causeIOException(e);
+		}
+	}
+
+	@Override
+	protected void handleClose() throws SailException {
+		contextIdIt.close();
+	}
+
+	protected SailException causeIOException(IOException e) {
```
Comment (hmottestad) [inline core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/LmdbContextIterator.java:60]: Is this meant to be overridden by subclasses?

File: core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/PersistentSetFactory.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+/**
+ * A factory for LMDB-based persistent sets.
+ */
+class PersistentSetFactory<T extends Serializable> {
+
+	final long env;
+	private final Path dbDir;
+	TxnManager txnManager;
+	long writeTxn;
+	PointerBuffer writeTxnPp = PointerBuffer.allocateDirect(1);
```
Comment (hmottestad) [inline core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/PersistentSetFactory.java:57]: Should any of the above variables be private?

**Review by hmottestad — COMMENTED** [review]
A bit too much for me to quite grasp how things work. I'll just assume that you know what your doing, it's your code after all ;)

File: core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/PersistentSetFactory.java (hunk 2 above)
Comment (kenwenzel) [reply core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/PersistentSetFactory.java:57]: I can't make them private for now as they are used as shared state by PersistentSet.
I could introduce getters/setters but they will slow down things (at least a bit).

File: core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/LmdbContextIterator.java (hunk 1 above)
Comment (kenwenzel) [reply core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/LmdbContextIterator.java:60]: No. I made the method private - also in LmdbStatementIterator.

**Comment by kenwenzel** [comment]
@hmottestad From my side this is now ready to merge.

**Pull request merged**